File: python/find_repeaters.py
import os, random, shutil, glob, time, crsmex, sys, smtplib
from subprocess import call
from datetime   import date
from obspy      import read
from obspy.core.util import gps2DistAzimuth

if len(sys.argv) == 1:  # UNIT TEST
	station_nm = "CAIG.HHZ..4-16Hz_data"
	send_email   = False
else:
	station_nm = str(sys.argv[1])

email_results = True	
share_results = True
root_dir      = "/home/u2/antonio/CRSMEX/data/"
path_data     = root_dir + station_nm + "/sac/"
coh_dir       = root_dir + station_nm + "/coh/"
log_dir       = root_dir + station_nm + "/log/"
out_dir       = "/home/u2/antonio/CRSMEX/output/"
share_dir     = "/data/seis01/antonio/output_crsmex/2015JUN10/"  

today         = date.today()
start_time    = time.time()
file_list     = sorted(glob.glob(os.path.join(path_data,"*.sac")))
N             = len(file_list)
out_file1     = "CRS_" + station_nm + "_" + str(today.year) + '{0:02d}'.format(today.month) + \
                                                              '{0:02d}'.format(today.day)   + ".DAT" # Luis's  format
out_file2     = "CRS_" + station_nm + "_" + str(today.year) + '{0:02d}'.format(today.month) + \
                                                              '{0:02d}'.format(today.day)   + ".BSL" # Taira's format
log_file      = "CRS_" + station_nm + "_" + str(today.year) + '{0:02d}'.format(today.month) + \
                                                              '{0:02d}'.format(today.day)   + ".LOG" # Log file 
out_file1     = out_dir + out_file1
out_file2     = out_dir + out_file2
log_file      = log_dir + log_file

# Control Parameters 
Win        =  25.5         # Win = 0 means dynamic window (see file eq_win.m), otherwise Win > 0 fixed lenght.
Threshold  =  0.6
low_f      =  1
high_f     =  8
n_poles    =  2
p_pick     = 'fixed'     # Manual a field, auto t5 field, combined  either a or t5 field
pplot      = True
no_limit   = True

eq_distance_threshold = 100
eq_sta_dist_limit     = 75

fid       = open(out_file1, "w")
fid2      = open(out_file2, "w")

computer  = os.popen("uname -m -n -s").read()
line_log  = "Windows width = " + str(Win) + " Threshold = " + str(Threshold) + " STNM = " + station_nm + \
           " Low = " + str(low_f) + "Hz. High = " + str(high_f) + "Hz. N_poles = " + str(n_poles) + " p_pick = " + p_pick + "."

fid.write("% " + computer)
fid.write("% Number of files: "    + str(N)   + "\n" )
fid.write("% "                     + line_log + "\n" )
fid.write("% Start Time = " + time.strftime("%d %B %Y at %H:%M:%S") + "\n")
print     "Searching repeaters for station " + station_nm + " ..."

for k in range(0,N):
    master = read( file_list[k] )
    master.filter('bandpass', freqmin=low_f , freqmax=high_f, corners=n_poles, zerophase=True )
    #master.filter('highpass', freq=low_f , corners=n_poles, zerophase=True )
    kevnm_master = master[0].stats.sac.kevnm.rstrip()
    log_line = 'k = ' + str(k) + ' out of ' + str(N)
    print log_line
    if (k-1)%100 == 0:
	fid_log   = open(log_file,  'a')
    	fid_log.write(log_line + '\n')
        fid_log.close()
    if no_limit == False:
    	eq_station = gps2DistAzimuth(master[0].stats.sac.evla, master[0].stats.sac.evlo, \
                                 master[0].stats.sac.stla, master[0].stats.sac.stlo)
    	if  master[0].stats.sac.dist > eq_sta_dist_limit:
		continue

    for n in range(k+1, N): 
        test         = read( file_list[n] )
        test.filter('bandpass', freqmin=low_f , freqmax=high_f, corners=n_poles, zerophase=True )
        #test.filter('highpass', freq=low_f ,  corners=n_poles, zerophase=True )
	eq_dist = 9999.  # Declare to a  large value
	if no_limit == False:
        	eq_dist = gps2DistAzimuth(master[0].stats.sac.evla, master[0].stats.sac.evlo, \
                                    test[0].stats.sac.evla,  test[0].stats.sac.evlo )[0]/1000
        if (eq_dist <= eq_distance_threshold) or (no_limit):
	    if p_pick == 'fixed':
		master_times = [master[0].stats.sac.t6, master[0].stats.sac.t7, master[0].stats.sac.t8]
		test_times   = [test[0].stats.sac.t6,   test[0].stats.sac.t7,   test[0].stats.sac.t8]
		for kfixed in range(0,3):
	   	    for jfixed in range(0,3):
		        p_master = master_times[kfixed]
		        p_test   = master_times[jfixed]
			if (p_master != -12345.) and (p_test != -12345.):
				CorrelationCoefficient, tshift, S1, S2 = crsmex.get_correlation_coefficient(master, test, Win, p_pick, pplot, p_master, p_test)
	    else:
            	CorrelationCoefficient, tshift, S1, S2 = crsmex.get_correlation_coefficient(master, test, Win, p_pick, pplot)
        else:
            continue
        if CorrelationCoefficient >= Threshold:
            fileout_coh = coh_dir + 'COH.' + 'BHZ.' + station_nm + '.' +  kevnm_master + '.' + test[0].stats.sac.kevnm.rstrip() + '.DAT'
            coherence   = crsmex.coherency(S1, S2, low_f, high_f, master[0].stats.sampling_rate, fileout_coh)
            outline     = str(k) + " " + file_list[k].split('/')[-1] + " " + file_list[n].split('/')[-1] + " " + \
                          "{0:6.4f}".format(CorrelationCoefficient) + " " \
		          "{0:6.2f}".format(tshift) + " " + "{0:5.2f}".format(master[0].stats.sac.evla) + " " + \
                          "{0:5.2f}".format(master[0].stats.sac.evlo) + " " + "{0:5.2f}".format(test[0].stats.sac.evla) + " " + \
                          "{0:5.2f}".format(test[0].stats.sac.evlo  ) + " " + "{0:5.2f}".format(Win) 
            outline2    = str(master[0].stats.starttime.year) + '.' + "{0:03d}".format(master[0].stats.starttime.julday) + '.' + \
                          "{0:02d}".format(master[0].stats.starttime.hour) + "{0:02d}".format(master[0].stats.starttime.minute) + \
                          "{0:02d}".format(master[0].stats.starttime.second) + ' ' + \
                          str(test[0].stats.starttime.year) + '.' + "{0:03d}".format(test[0].stats.starttime.julday) + '.' + \
                          "{0:02d}".format(test[0].stats.starttime.hour) + "{0:02d}".format(test[0].stats.starttime.minute) + \
                          "{0:02d}".format(test[0].stats.starttime.second) + ' ' + \
                          "{0:04d}".format(int(CorrelationCoefficient*1e4)) + ' ' + "{0:04d}".format(int(coherence*1e4)) + ' ' + \
			  master[0].stats.sac.kevnm.rstrip() + ' ' + test[0].stats.sac.kevnm.rstrip() 
            print outline2
	    fid.write(outline + '\n')
	    fid2.write(outline2 + '\n')
	    outline=''
            outline2=''
print "Writting ", out_file1
print "Writting ", out_file2
print "Elapse time (minutes) = ", (time.time() - start_time)/60., ", (hours) = ", (time.time() - start_time)/3600.
fid.write("% End Time = " + time.strftime("%d %B %Y at %H:%M:%S") + "\n")
fid.close()
fid2.close()
fid_log.close()

if share_results:
	shutil.copy(out_file1, share_dir)
	shutil.copy(out_file2, share_dir)
        
if email_results:
	FROM    = 'antonio@seismo.berkeley.edu'
	TO      = 'antonio@moho.ess.ucla.edu'
	SUBJECT = '[CRSMEX] Results from station ' + station_nm
	TEXT    = line_log
	message = 'Subject: %s\n\n%s' % (SUBJECT, TEXT)

	server = smtplib.SMTP('mail.geo.berkeley.edu')
	server.sendmail(FROM, TO, message)
	server.quit()        

File: matlab/find_repeaters.m
function find_repeaters(station_nm)
tic;
close all
clc

if nargin == 0
    station_nm = 'CAIG.HHZ..4-16Hz_data';
end

home_dir = '/home/u2/antonio/';
root_dir = '/home/u2/antonio/CRSMEX/';
email_results = 1;

addpath([home_dir 'CJA_Msac1.1/io'])
addpath([home_dir 'lib'])
addpath([root_dir 'matlab'])

%station_nm = 'CAIG';
BTime = datestr(now);
ddate = date;
ddate = [ddate(1:2) ddate(4:6) ddate(8:11)];

Win        =  40.0;    % Win = 0 means dynamic window (see file eq_win.m), otherwise Win > 0 fixed lenght.
Threshold  =   0.6;
low_f      =   1.0;
high_f     =   8.0;
n_poles    =   4;
p_pick     = 'combined';     % manual a field, auto t5 field, combined  either a or t5 field
eq_distance_threshold = 100;

cd([root_dir 'data/' station_nm '/sac']);
inputfiles = [root_dir 'data/' station_nm '/sac/list_all.info'];
out_file1  = ['CRS_' station_nm '_' BTime(8:11) upper(BTime(4:6)) BTime(1:2) '.DAT'];
out_file2  = ['CRS_' station_nm '_' BTime(8:11) upper(BTime(4:6)) BTime(1:2) '.BSL']; % Compatible with Taka'aki's routines
log_file   = ['CRS_' station_nm '_' BTime(8:11) upper(BTime(4:6)) BTime(1:2) '.LOG'];

fid        = fopen(out_file1,  'w');
fid2       = fopen(out_file2, 'w');
fid_log    = fopen(log_file,  'w');

A          = importdata(inputfiles);
N          = numel(A);

[dummy comp_info]  = system('uname -m -n -s');
line_log   = ['% Windows width = ' num2str(Win) ' Threshold = ' num2str(Threshold) ' STNM = ' station_nm...
              ' Low = ' num2str(low_f) 'Hz. High = ' num2str(high_f) 'Hz. N_poles = ' num2str(n_poles)  'p_pick = ' p_pick '.'];
fprintf(fid,  '% s',['% ' comp_info]);
fprintf(fid,  '% s\n',['% ' inputfiles]);
fprintf(fid,  '% Number of files: %d\n',N);
fprintf(fid,  '% Data proccesed on %s\n',date);
fprintf(fid,  '%s\n', line_log);
fprintf(fid,  '%s\n', ['% Start Time = ' BTime ]);

disp(['Searching repeaters for station ' station_nm ' ...'])

for k = 1:N
    master  = rsac(fullfile(pwd,char(A(k))));
    master  = filter_sac(master,low_f, high_f, n_poles); 
    fsample = 1/master.dt; 
    for n = k+1:N
        test    = rsac(fullfile(pwd,char(A(n))));
        disp([num2str([1/test.dt]) ' ' test.filename])
        test    = filter_sac(test, low_f, high_f, n_poles);
        eq_dist = distkm([test.evla test.evlo],[master.evla master.evlo]);
        if eq_dist <= eq_distance_threshold           
            [CorrelationCoefficient tshift S1 S2] = get_correlation_coefficient(master, test, Win, p_pick);
        else
            continue;
        end
        if CorrelationCoefficient >= Threshold
	    coherence  = coherency(S1, S2, low_f, high_f, fsample);
            outline = [ num2str(k) ' ' char(A(k)) ' ' char(A(n)) ' ' ...
                        num2str(CorrelationCoefficient,'%6.4f')  ' ' ...
                        num2str(tshift,'%6.2f') ' ' num2str(master.evla,'%5.2f') ' ' ...
                        num2str(master.evlo,'%6.2f') ' '  num2str(test.evla,'%5.2f') ];
            outline2 = [num2str(master.nz(1),'%.4d')  '.' num2str(master.nz(2),'%.3d') '.' ...
                        num2str(master.nz(3),'%.2d')      num2str(master.nz(4),'%.2d') ...
                        num2str(master.nz(5),'%.2d') ' '                               ...
                        num2str(test.nz(1),'%.4d')    '.' num2str(test.nz(2),'%.3d')   '.' ...
                        num2str(test.nz(3),'%.2d')        num2str(test.nz(4),'%.2d')  ...
                        num2str(test.nz(5),'%.2d')  ' ' ...
                        num2str(round(CorrelationCoefficient*1e4)) ' ' num2str(round(coherence*1e4)) ' ' master.kevnm ' ' test.kevnm];
            disp(outline)
            fprintf(fid, '%s\n',outline );
	        fprintf(fid2,'%s\n',outline2);
        end
    end
    log_line = ['k = ' num2str(k) ' out of ' num2str(N)];
    disp(log_line)
    fprintf(fid_log,'%s\n',log_line);
end
elapsed_time=toc/3600;   % Computing time in hours
ETime = datestr(now);
fprintf(fid, '%s\n', ['% End Time = ' ETime ' - Elapse Time = ' num2str(elapsed_time)]);
fclose(fid);
fclose(fid2);
fclose(fid_log);

disp(BTime)
disp(ETime)
system(['chmod -w ' out_file1]);
system(['chmod -w ' out_file2]);
system(['cp ' out_file2 ' ' home_dir 'output_crsmex/'])

if email_results == 1
	subject = ['[CRSMEX] Results from station ' station_nm '.'];
        message = line_log;
	setpref('Internet','E_mail','antonio@seismo.berkeley.edu');
	setpref('Internet','SMTP_Server','mail.geo.berkeley.edu')
	sendmail('antonio@moho.ess.ucla.edu',subject, message, {out_file1,out_file2,})
end
